### ISSUE TITLE
Improved Embed Security

### ISSUE THREAD

**Issue by MarcGJA** [metabase/metabase#8111]
The primary concern is locking down a secure embed's iframe url so that it cannot be extracted and shared among app users with different access levels.

Take, for example, an app featuring Metabase secure embeds for 3 levels of increasing BI access.  At present there doesn't appear to be a way to prevent users with the highest level of access from viewing the secure embed's iframe url via some browser tool and sharing it among those with a lower level of access.  While <b><a href="https://www.metabase.com/docs/latest/administration-guide/10-single-sign-on.html">Authenticating with Google Sign-In or LDAP</a></b> and <b><a href="https://metabase.com/docs/latest/administration-guide/13-embedding.html">locked parameters</a></b> are excellent optional Metabase features they do not prevent the sharing of token-ized iframe src urls.  

<b>A couple of possible solutions</b>

<b>JWT re-issuing</b> with every filter selection within the secure embed's iframe allowing for very short-lived expiring tokens (seconds at most).  Expiring JWTs (15m, 1hr, etc.) using iat, nbf, exp claims is a very good practice for any app developer to implement on their own outside of Metabase (see below).  But for however long the token is valid is however long the secure embed's url is exposed to exploitation.  Were it possible latch onto filter selection changes and other actions within a secure embed's iframe (see <a href="https://github.com/metabase/metabase/pull/7817">Pull Request #7817</a>) it might just be possible to implement very short-lived JWTs solving this issue completely.

<b>Permit more than one sub-domain</b> via the Admin>General>Site URL field (e.g. http:// metabase<b>01</b>.my-company.com, http:// metabase<b>02</b>.my-company.com, http:// metabase<b>03</b>.my-company.com), one per access level.  Presumably secure embed urls generated for each access level would be rendered useless unless logged into the correct sub-domain.

<b>Other possibilities that do not appear viable</b>
 
<b>Encryption</b> is a desirable enhancement (see <a href="https://github.com/metabase/metabase/pull/4271#issuecomment-286419875">Pull Request #4271 comment</a>) but does not prevent the sharing of token-ized iframe src urls.   

<b>Single Sign-on Embedding</b> is more in line with a hosted service is it not?  Perhaps at some later date Metabase may offer such a service but until then SSO is probably out.

<b>Obfuscation, substitution, or otherwise hiding</b> html code is ineffective as at some point the secure embed's url must be readable for use.

PS -- JWT expiration is shown below and modeled after the <a href="https://github.com/metabase/embedding-reference-apps/tree/master/laravel/embedded_analytics">Laravel (PHP)</a> Metabase example reference app using the <a href="https://github.com/lcobucci/jwt">lcobucci/jwt</a> library:

    $metabaseSiteUrl = 'https:// metabase.my-company.com/'; // Url of the Metabase installation (same as Admin>General>Site URL field)
    $metabaseSecretKey = '08c0a...6b3ec'; // Secret key generated by Metabase installation

    $signer = new Sha256();
    $token = (new Builder())

        ->setIssuedAt(time()) // Configures the time that the token was issue (iat claim)
        ->setNotBefore(time() + 60) // Configures the time that the token can be used (nbf claim)
        ->setExpiration(time() + 3600) // Configures the expiration time of the token (exp claim)

        ->set('resource', [
            'dashboard' => 1 // ID of the card or dashboard resource to implement
        ])
        ->set('params', [
            'params' => [] // Corresponds to Metabase locked parameters (e.g. userid or some other filterable data category that can serve as a locked parameter)
        ])
        ->sign($signer, $metabaseSecretKey)
        ->getToken();

**Comment by nitzien** [comment]
Signature not expiring is serious security issue. Any idea when fix for this is possible? Also, in this case we will need to refresh whole embedding page for csv /xlsx download if download link has expired.

**Comment by MarcGJA** [comment]
@nitzien, I'm not sure the exact nature of your concerns but I'll do my best.  

Expiring signatures for embeds is a good thing and you can just leave it out of the picture entirely if you like.  I wouldn't and I won't because if you don't make sure to have written expiration into the code yourself the secure embeds you create are basically good forever making for a security issue in many cases.  

Secondly that particular aspect is NOT a Metabase issue or "fix", that's the responsibility of the developer. Metabase just provides some helpful examples of what the code can look like and expiration isn't specifically addressed.  

Thirdly, my issue came up when I suddenly realized that the tokenized iframe url once accessed by a bad actor can be copy-pasted into a new window or shared.  That's when I felt the need to take steps and expiration is one of those steps.  But yes once the link is expired a user needs go back to whatever mechanism you as the developer have put into place to select the resource again and have it come up fresh with a new token.  Thing is if the user had all their filters setup and had it all the way they wanted it they've got set that all up all over again.  As far as I know Metabase does not as yet allow the developer a means of "listening-in" on the contents of the iframe on the client side in such a way as to permit the client-side user of embeds the ability to save off their favorite filter setups for a given resource.  At least one Metabase developer has gone thru the trouble of creating a whole separate mechanism within their own application to handle the filters natively and feed the selections to the Metabase embeds so that way their client side users can save off their favorite filter setups for use anytime... Don't ask me how it works, it sounds intriguing, but the time, the effort, the maintenance of such a thing is just not a palatable solution for me right now... I just want something basic off the ground ASAP and we'll see if Metabase eventually gets around to that <a href="https://github.com/metabase/metabase/pull/7817">Pull Request #7817</a> and if that provides a more streamlined way of doing it.

Bottom line, I want to lock down that iframed tokenized url but I don't know if it is even possible.  I simply do not have the expertise to address this issue, all I can do is make the concern known, toss out a few ideas, and do my best to limit access to those urls any way I can in my own app.  But definitely, if you're dealing with sensitive information and want to do embedding be careful!  You're going outside of the Metabase application and it's now your responsibility to keep things secure!  And if the Metabase team or community can help with that, give helpful suggestions, fantastic, any assistance would be welcome!

**Comment by nitzien** [comment]
Hey @MarcGJA, I am sorry if I am misunderstood. I meant same as what you did. Your third point is exactly my concern. Ideal case will be to use exp, nbf and iat (ttps://tools.ietf.org/html/rfc7519#section-4.1) in jwt. Exp can be set as session expiry time added to current time. This will give better security.  I believe this will need fix from metabase.

**Comment by tlrobinson** [comment]
> Metabase just provides some helpful examples of what the code can look like and expiration isn't specifically addressed.

This is definitely something we should do. In the original PR I suggested we at least include expirations in the sample code, and possibly require it (you could set it to a far future date if you really wanted to)

> As far as I know Metabase does not as yet allow the developer a means of "listening-in" on the contents of the iframe on the client side in such a way as to permit the client-side user of embeds the ability to save off their favorite filter setups for a given resource. 

That is correct. You could potentially control the parameters using your own UI (generating a new token each time they're changed) rather than using Metabase's parameters UI. However it looks like we currently don't allow optional signed parameters so that may not work depending on what you're trying to do.

**Comment by JulianNorton** [comment]
For what it's worth, here is how Google addresses it with their APIs:

![screen shot 2018-11-30 at 2 25 56 pm](https://user-images.githubusercontent.com/2412992/49312772-2e6cb680-f4b3-11e8-95c4-767b6d28d728.png)


I think Metabase could implement a similar feature where only certain referrer IP addresses would be allowed.

**Comment by snarfed** [comment]
> As far as I know Metabase does not as yet allow the developer a means of "listening-in" on the contents of the iframe on the client side in such a way as to permit the client-side user of embeds the ability to save off their favorite filter setups for a given resource.

@MarcGJA @tlrobinson just fyi this is why i did #7817. it makes embedded dashboards send their parameter values up to the parent document in a JS event. it also shows how the parent doc can propagate parameter values back down into the dashboard by changing the iframe's URL.

you can see this in action on https://data.color.com/v1/ . if you change a filter, the top level browser URL updates accordingly. you can then open that URL in a new tab and it will use the same parameter values.

**Comment by tlrobinson** [comment]
Anyone interested in this issue may be interested in my proposal in #9494

**Comment by tlrobinson** [comment]
@MarcGJA I think supporting re-issuing of tokens could allow for two different (but not mutually exclusive) features:

1. locking a token to a specific set of parameter values, but still allowing the user to change the parameter, which would trigger a request to re-issue the token (this would also allow the embedding application fine control over which parameter values are allowed by each user)
2. expiring the token very quickly, so it can't be easily shared

As you mentioned, the former can be implemented using your own parameter "chrome" and something like @snarfed's #7817 (I also have something similar in a private repo I'll try to extract). It would be nice you didn't have to re-implement the parameter chrome, though.

I left some thoughts about this at the bottom of #9494, but I'm curious if you have ideas about how to best implement token reissuing. Is there any prior art in similar projects you're aware of?

**Comment by MarcGJA** [comment]
@tlrobinson On JWT re-issuing, I'm not aware of any examples out there, surely they do exist.  From <a href="https://github.com/metabase/metabase/issues/9494">#9494</a> it sounds like it's doable.  

Ideally the end user wouldn't even notice as they change filter parameters throughout a session.  And security would be greatly improved as copy-pasting the "token-ized" URL from one browser address bar to another or emailing as a link would fail because the continuously re-issued JWTs behind it expire almost immediately.  Between the Metabase team and all the talented programmers in this forum a solution will be worked out.

**Comment by MyckeTap** [comment]
are there any near future plans to improve security in the given context?

**Comment by mmoole** [comment]
here are some more ideas for handling the token server side in order to improve security a bit:
* restrict the token to be used only for
  * the first remote host (ip address for example), then don't allow other hosts use this token
  * the first user agent, then don't allow other user agents use this token
  * the first usage - the token is seen as expired after that by the server
* expire the token by defining the timeframe for this on the server instead of including it in the token
* use a JTI claim for the JWT 
* generate and handle JWTs with ['selective disclosure'](https://tools.ietf.org/id/draft-fett-oauth-selective-disclosure-jwt-02.html)
